src/charm.py

```diff
@@ -133,32 +134,39 @@ def _configure(self, _):
         # Restart prometheus only if command line arguments have changed,
         # otherwise just reload its configuration.
         if current_services == new_layer.services:
+            # No change in layer; reload config to make sure it is valid
             external_url = urlparse(self._external_url)
-
             prometheus_server = Prometheus(web_route_prefix=external_url.path)
-
             reloaded = prometheus_server.reload_configuration()
             if not reloaded:
                 logger.error("Prometheus failed to reload the configuration")
                 self.unit.status = BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)
                 return
+
             logger.info("Prometheus configuration reloaded")
+
         else:
+            # Layer changed - replan.
             container.add_layer(self._name, new_layer, combine=True)
             try:
+                # If a config is invalid then prometheus would exist immediately.
+                # This would be caught by pebble (default timeout is 30 sec) and a ChangeError
+                # would be raised.
                 container.replan()
                 logger.info("Prometheus (re)started")
-            except Exception as e:
-                logger.error(f"Failed to replan; pebble plan: {container.get_plan().to_dict()}")
-                raise e
-
-        if (
-            isinstance(self.unit.status, BlockedStatus)
-            and self.unit.status.message != CORRUPT_PROMETHEUS_CONFIG_MESSAGE
-        ):
-            return
+            except ChangeError as e:
+                logger.error(
+                    "Failed to replan; pebble plan: %s; %s", container.get_plan().to_dict(), str(e)
+                )
+                self.unit.status = BlockedStatus(REPLAN_FAILED_MESSAGE)
+                return
 
-        self.unit.status = ActiveStatus()
```

Comment:
@balbirthomas iiuc, the existing code would incorrectly set the status to Active if the current status is e.g. Maintenance.